### ISSUE TITLE
Feature request: Implement for loops with a for iterators

### ISSUE THREAD

**Issue by 64kramsystem** [immunant/c2rust#621]
For loops are currently transpiled to while loops. I don't know if this is necessary in some cases, however, in the cases I've dealt with, the `for` loop could be converted directly, which makes the output considerably cleaner.

Example (simplified from the source):

```c
for (loop = 0; loop < 16; loop++)
    if (digit == hexstr[loop]) {
      break;
    }
}
```

```rs
loop_0 = 0;
while loop_0 < 16 {
    if digit as i32 == hexstr[loop_0 as usize] as i32 {
        break;
    } else {
        loop_0 = loop_0.wrapping_add(1);
    }
}
```

Based on my understanding, C2Rust is trying to generalize a for loop where there can be a wraparound, however, there is an upper bound here, so there's no need for specific handling.

The simpler version is:

```rs
for loop_0 in 0..16 {
    if digit == hexstr[loop_0 as usize] {
        break;
    }
}
```

**Comment by kkysen** [comment]
You're right that because the loop here has a constant bound, it can be simplified, but obviously that's not true in the general case.  I'm not sure how difficult it is in the transpiler to detect such a thing, though.  However, I do think the very common C pattern of

```c
for (T i = start; i < end; i++) {
    ....
}
```

for integer types `T` where `start` and `end` are also of type `T` should be able to be converted to

```rust
for i in start..end {
    ...
}
```

because the upper bound (whether constant or not) ensures there will be no overflow.  We'd just need to check that `start` and `end` are the same type, as otherwise casts and promotions could allow an overflow.

**Comment by Rua** [comment]
Something for a future implementer to take note of: you also need to check that none of the variables mentioned in the `for` header are modified inside the loop.
